plugins/channel-safely.cpp

```diff
@@ -0,0 +1,480 @@
+
+DECLARE_HASA(when) //declares above statement with 'when' replacing 'what'
+// end usage is: `has_when<T>::value`
+// the only use is to allow reliance on pull request #1876 which introduces a refactor which prevents some manual management
+
+ChannelManager manager;
+void onNewHour(color_ostream &out, void* tick_ptr);
+void onNewDay(color_ostream &out, void* tick_ptr);
+void onStart(color_ostream &out, void* job);
+void onComplete(color_ostream &out, void* job);
+
+void manageNeighbours(color_ostream &out, const df::coord &tile);
+void cancelJob(df::job* job);
+bool is_group_done(const GroupData::Group &group);
+bool is_dig(df::job* job);
+bool is_channel(df::job* job);
+bool is_channel(df::tile_designation &designation);
+
+color_ostream* debug_out = nullptr;
+
+command_result manage_channel_designations (color_ostream &out, std::vector <std::string> &parameters);
+
+DFhackCExport command_result plugin_init( color_ostream &out, std::vector<PluginCommand> &commands) {
+    commands.push_back(PluginCommand("channel-safely",
+                                     "A tool to manage active channel designations.",
+                                     manage_channel_designations,
+                                     false,
+                                     "\n"));
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    namespace EM = EventManager;
+    if(enable && !enabled) {
+        using namespace EM::EventType;
+        EM::EventHandler hoursHandler(onNewHour, hourTicks);
+        EM::EventHandler daysHandler(onNewDay, dayTicks);
+        EM::EventHandler jobStartHandler(onStart, 0);
+        EM::EventHandler jobCompletionHandler(onComplete, 0);
+
+        if(!has_when<EM::EventHandler>::value){
+            EM::registerTick(hoursHandler, hourTicks, plugin_self);
+            EM::registerTick(daysHandler, dayTicks, plugin_self);
+        } else {
+            EM::registerListener(EventType::TICK, hoursHandler, plugin_self);
+            EM::registerListener(EventType::TICK, daysHandler, plugin_self);
+        }
+        EM::registerListener(EventType::JOB_INITIATED, jobStartHandler, plugin_self);
+        EM::registerListener(EventType::JOB_COMPLETED, jobCompletionHandler, plugin_self);
+        manager.manage_designations(out, true);
+        out.print("channel-safely enabled!\n");
+    } else if (!enable) {
+        EM::unregisterAll(plugin_self);
+        out.print("channel-safely disabled!\n");
+    }
+    enabled = enable;
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+    if(enabled && World::isFortressMode()) {
+        switch (event) {
+            case SC_UNKNOWN:
+                break;
+            case SC_WORLD_LOADED:
+                break;
+            case SC_WORLD_UNLOADED:
+                break;
+            case SC_MAP_LOADED:
+                manager.manage_designations(out, true);
```

Comment:
If it's only going to handle a single event type, this `switch` statement can probably be collapsed into a single `if` statement.

Reply:
I figured the compiler would do that for me